keps/sig-node/2727-grpc-probe/README.md

```diff
@@ -0,0 +1,177 @@
+    - [GA](#ga)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+- [References](#references)
+<!-- /toc -->
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in
+  [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture
+      and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in
+  [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents,
+  links to mailing list discussions/SIG meetings, relevant PRs/issues,
+  release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Goals
+
+Enable gRPC probe natively from Kubelet without requiring users to package a
+gRPC healthcheck binary with their container.
+
+- https://github.com/grpc-ecosystem/grpc-health-probe
+- https://github.com/grpc/grpc/blob/master/doc/health-checking.md
+
+## Non-Goals
+
+Add gRPC support in other areas of K8s (e.g. Services).
+
+## Proposal
+
+Add the follow configuration to the `LivenessProbe`, `ReadinessProbe`
+and `StartupProbe`. Example:
+
+```yaml
+    readinessProbe:
+      grpc:                     #+
+        port: 9090              #+
+        service: my-service     #+
+      initialDelaySeconds: 5
+      periodSeconds: 10
+```
+
+This will result in the use of gRPC (using HTTP/2) to use the standard
+healthcheck service to determine the health of the container. We do
+not support other protocols for the time being (HTTP, QUIC), those can
+be enabled in Beta with user feedback.
+
+The healthcheck request will be identified with the following gRPC
+`User-Agent` metadata. This user agent will be statically defined (not
+configurable by the user):
+
+```
+User-Agent: kubernetes/K8S_MAJOR_VER.K8S_MINOR_VER
+```
+
+Example:
+
+```
+User-Agent: kubernetes/1.22
+```
+
+### Risks and Mitigations
+
+1. Adds more code to Kubelet and surface area to Pod.Spec. *Response*: we
+   expect that this will be generally useful given broad gRPC adoption in the
+   industry.
+
+## Design Details
+
+```go
+// core/v1/types.go
+
+type Handler struct {
+  // ...
+  TCPSocket *TCPSocketAction `json...`
+
+  // GRPC specifies an action involving a TCP port.          //+
+  // +optional                                               //+
+  GRPC *GRPCAction `json...`                                 //+
+
+  // ...
+}
+
+type GRPCAction struct {
```

Comment:
to which IP is connecting it?
the TCP and HTTP probeActions have
```
	// Optional: Host name to connect to, defaults to the pod IP.
	// +optional
	Host string
```

Reply:
That actually sounds kind of dangerous -- can you use this to have kubelet ping arbitrary hosts?

Reply:
https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#:~:text=here's%20one%20scenario%20where%20you%20would%20set%20it.%20suppose%20the%20container%20listens%20on%20127.0.0.1%20and%20the%20pod's%20hostnetwork%20field%20is%20true.%20then%20host%2C%20under%20httpget%2C%20should%20be%20set%20to%20127.0.0.1.

> Here's one scenario where you would set it. Suppose the container listens on 127.0.0.1 and the Pod's hostNetwork field is true. Then host, under httpGet, should be set to 127.0.0.1.

Reply:
heh, I know, that is the problem with the other probes, but I don't know if there is some compatibility we want to keep

Reply:
it is more a general question, I prefer probe to  mean "probe the pod not something else" but maybe there was some historical reasons

Reply:
damn latency, Sergey already answered :)

Reply:
Thought about it, adding validations is that for a hostNetwork Pod, the IP could literally be anything.
Minor exfil route I guess, but probably not very critical.

Reply:
Added Host.